lib/index.ts

```diff
@@ -0,0 +1,572 @@
+type InternalCallback = (err: Error | null, response: any | null) => void;
+
+export class Front {
+	public comment: {
+		create: (params: FrontCommentRequest.Create, callback?: FrontCallback<FrontComment>) => Promise<FrontComment>;
+		get: (params: FrontCommentRequest.Get, callback?: FrontCallback<FrontComment>) => Promise<FrontComment>;
+		listMentions: (params: FrontCommentRequest.ListMentions, callback?: FrontCallback<FrontCommentMentions>)
+			=> Promise<FrontCommentMentions>;
+	};
+
+	public conversation: {
+		get: (params: FrontConversationRequest.Get, callback?: FrontCallback<FrontConversation>)
+			=> Promise<FrontConversation>;
+		list: (params?: FrontConversationRequest.List, callback?: FrontCallback<FrontConversations>)
+			=> Promise<FrontConversations>;
+		listComments: (params: FrontConversationRequest.ListComments, callback?: FrontCallback<FrontConversationComments>)
+			=> Promise<FrontConversationComments>;
+		listFollowers: (params: FrontConversationRequest.ListFollowers,
+			callback?: FrontCallback<FrontConversationFollowers>) => Promise<FrontConversationFollowers>;
+		listInboxes: (params: FrontConversationRequest.ListInboxes,
+			callback?: FrontCallback<FrontConversationInboxes>) => Promise<FrontConversationInboxes>;
+		listMessages: (params: FrontConversationRequest.ListMessages,
+			callback?: FrontCallback<FrontConversationMessages>) => Promise<FrontConversationMessages>;
+		update: (params: FrontConversationRequest.Update, callback?: FrontCallback<void>) => Promise<void>;
+	};
+
+	public inbox: {
+		create: (params: FrontInboxRequest.Create, callback?: FrontCallback<FrontInboxCreation>)
+			=> Promise<FrontInboxCreation>;
+		get: (params: FrontInboxRequest.Get, callback?: FrontCallback<FrontInbox>) => Promise<FrontInbox>;
+		list: (callback?: FrontCallback<FrontInboxes>) => Promise<FrontInboxes>;
+		listChannels: (params: FrontInboxRequest.ListChannels, callback?: FrontCallback<FrontInboxChannels>)
+			=> Promise<FrontInboxChannels>;
+		listConversations: (params: FrontInboxRequest.ListConversations,
+			callback?: FrontCallback<FrontInboxConversations>) => Promise<FrontInboxConversations>;
+		listTeammates: (params: FrontInboxRequest.ListTeammates, callback?: FrontCallback<FrontInboxTeammates>)
+			=> Promise<FrontInboxTeammates>;
+	};
+
+	public message: {
+		get: (params: FrontMessageRequest.Get, callback?: FrontCallback<FrontMessage>) => Promise<FrontMessage>;
+		send: (params: FrontMessageRequest.Send, callback?: FrontCallback<FrontConversationReference>)
+			=> Promise<FrontConversationReference>;
+		reply: (params: FrontMessageRequest.Reply, callback?: FrontCallback<FrontStatus>) => Promise<FrontStatus>;
+		receiveCustom: (params: FrontMessageRequest.ReceiveCustom,
+			callback?: FrontCallback<FrontConversationReference>) => Promise<FrontConversationReference>;
+	};
+
+	private apiKey: string;
+
+	constructor(apiKey: string) {
+		// Key.
+		this.apiKey = apiKey;
+
+		// Setup resource paths.
+		this.comment = {
```

Comment:
But the problem with this is that it's not the right scope? `Front` is the scope we want, but in fact the scope differs because of the sub-scoped objects (`inbox`, `message`, `conversation`, etc).

eg., using your comment, what I want is something like:
```typescript
public inbox = {
    create: function(params: InboxRequest.Create, callback?: Callback<InboxCreation>): Promise<InboxCreation> {
         return this.httpCall.apply(this, [{ reqType: 'POST', path: 'conversations/{conversation_id}/comments' }].concat(arguments))
    }
}
```
but bound to the Front and not the `inbox` scope (which is why I defined them separately then instantiated them in the constructor). 

Sorry @Page- , I'm probably being really dim here.